lib/matchers/matchers.rb

```diff
@@ -338,3 +338,164 @@ def format_expectation(negate)
+#
+# TODO explain somewhere that :all_with_args, :all_without_args, :all_with_integer_arg
+# will cause match_pam_rule to return true when there are no potential matches
+RSpec::Matchers.define :match_pam_rule do |expected|
+  def matching_integer_arg? (line)
+    line.module_arguments.any? do |arg|
+      key, value = arg.split('=')
+
+      value && (@args[:key] == key) && value.match?(/^-?\d+$/) &&
+        value.to_i.send(@args[:operator].to_sym, @args[:value])
+    end
+  end
+
+  match do |actual|
+    case @args_type
+    when :all_with_args, :all_without_args, :all_with_integer_arg
+      retval = true
+    when :any_with_args, :any_with_integer_arg
+      retval = false
+    end
+
+    if [:all_with_integer_arg, :any_with_integer_arg].include? @args_type
+      unless Numeric.method_defined?(@args[:operator])
+        fail("Error: Operator '#{@args[:operator]}' is an invalid numeric comparison operator.")
+      end
+    end
+
+    actual_munge = {}
+
+    @expected = expected.to_s
+
+    if @args_type
+      catch :stop_searching do
+        actual.services.each do |service|
+          expected_line = Pam::Rule.new(expected, {:service_name => service})
+
+          potentials = actual.find_all do |line|
+            line.match?(expected_line)
+          end
+
+          if potentials && !potentials.empty?
+            actual_munge[service] ||= []
+            actual_munge[service] += potentials.map(&:to_s)
+
+            potentials.each do |potential|
+              case @args_type
+              when :all_without_args
+                retval = !potential.module_arguments.join(' ').match?(@args)
+                throw :stop_searching unless retval
+              when :all_with_args
+                retval = potential.module_arguments.join(' ').match?(@args)
+                throw :stop_searching unless retval
+              when :all_with_integer_arg
+                retval = matching_integer_arg? potential
+                throw :stop_searching unless retval
+              when :any_with_integer_arg
+                retval = matching_integer_arg? potential
+                throw :stop_searching if retval
+              when :any_with_args
+                retval = potential.module_arguments.join(' ').match?(@args)
+                throw :stop_searching if retval
+              end
+            end
+          end
+        end
+      end
+    else
+      retval = actual.include?(expected, {:service_name => actual.service})
+    end
+
+    if actual_munge.empty?
+      @actual = actual.to_s
+    elsif actual_munge.keys.length == 1
+      @actual = actual_munge.values.flatten.join("\n")
+    else
+      @actual = actual_munge.map do |service, lines|
+        lines.map do |line|
+          service + ' ' + line
+        end
+      end.flatten.join("\n")
+    end
+
+    retval
+  end
+
+  diffable
+
+  # TODO make these an array of args so that we can actually chain them together
+  chain :any_with_args do |args|
```

Comment:
Overall, I understand the desire to make a matcher that is very RSPec-like.  However, we generally don't expect the user to know RSpec at all, and we have found that chaining matchers is harder to write compared to having matchers that take arguments.

This isn't a blocker for me, but it would have to be well-documented.

Reply:
Honestly, this just read better and required less code to figure out in the tests.

I get that users shouldn't have to know rspec but, at a certain point, they're going to have to if they want to write tests that aren't insane levels of weird BASH-fu.